Generalize service implementations

darinkrauss opened the pull request:
- Implement general ServicesManager and service serialization/restoration
- Replace OSLog and DiagnosticLogger with DiagnosticLog and connect to LoggingManager
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation to LoopKit
- Move NightscoutUploader to NightscoutService
- Move Amplitude, Loggly, and Nightscout service implementations to respective repos
- Implement generalized LoggingManager, AnalyticsManager, and RemoteDataManager to connect to specific services
- Refactor Analytics singleton to injectable implementation
- Remove unused MLabService KeyChain extension
- Refactor SettingsTableViewController to generalize service setup

https://tidepool.atlassian.net/browse/LOOP-11
https://tidepool.atlassian.net/browse/LOOP-13
https://tidepool.atlassian.net/browse/LOOP-10
https://tidepool.atlassian.net/browse/LOOP-27
https://tidepool.atlassian.net/browse/LOOP-22

darinkrauss committed:
Generalize service implementations

- Implement general ServicesManager and service serialization/restoration
- Replace OSLog and DiagnosticLogger with DiagnosticLog and connect to LoggingManager
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation to LoopKit
- Move NightscoutUploader to NightscoutService
- Move Amplitude, Loggly, and Nightscout service implementations to respective repos
- Implement generalized LoggingManager, AnalyticsManager, and RemoteDataManager to connect to specific services
- Refactor Analytics singleton to injectable implementation
- Remove unused MLabService KeyChain extension
- Refactor SettingsTableViewController to generalize service setup

nhamming commented on Loop/AppDelegate.swift:
```diff
@@ -8,28 +8,41 @@
+
+    private lazy var analyticsManager = AnalyticsManager(servicesManager: servicesManager)
+
+    private lazy var loggingManager = LoggingManager(servicesManager: servicesManager)
```
Should some of these be constants?

darinkrauss replied:
I believe a `lazy` property must be a `var` (not `let`).

nhamming commented on Loop/Managers/RemoteDataManager.swift:
```diff
@@ -8,34 +8,177 @@
+        // TODO: How to handle completion correctly
+        if remoteData.count > 0 {
+            remoteData[0].upload(pumpEvents: events, fromSource: source, completion: completion)
+        }
```
Why only the first element? Similar comment for code below.

darinkrauss replied:
This is a placeholder that will only work for one `RemoteData` service at a time. However, at this point with the code there is only one `RemoteData` service anyways, namely `NightscoutService`. 

This is one reason why we are moving from the current "push" model to a "pull" model. The "pull" model will allow multiple `RemoteData` services to be active at the same time (eg. `NightscoutService` and `TidepoolService`).

So, this is a "good enough for now" situation that will be updated with the next round of updates to the `RemoteData` service and `NightscoutService`.

nhamming commented on Loop/Managers/Service.swift:
```diff
@@ -0,0 +1,51 @@
+    NightscoutService.self,
+    LogglyService.self,
+    AmplitudeService.self,
+    MockService.self,
```
Should the MockService be limited to just debug builds?

darinkrauss replied:
Yes. Will make that change.

nhamming reviewed (approved):
That is all.

darinkrauss committed:
Updates from PR review

- Update MockService to only be included in debug builds
- Move shared logging setup to application didFinishLaunchingWithOptions
- Fix issues with service rows in settings view
- Fix build issues with copying service frameworks

darinkrauss committed:
Updates reflecting service protocol renaming

- Update service protocol usage based upon renaming
- Rename AnalyticsManager to AnalyticsServicesManager
- Rename LoggingManager to LoggingServicesManager
- Rename RemoteDataManager to RemoteDataServicesManager

darinkrauss committed:
Update service completion and notification delegate from PR review

darinkrauss committed:
Remove extraneous whitespace

nhamming reviewed (approved):
All good!

darinkrauss committed:
Refactor service setup delegate from PR comments

darinkrauss committed:
Remove unnecessary elements of Service protocol

ps2 reviewed (approved):
🚀 Looks good!
